File: dev_tools/ehp_parsing/changelog.py
import json
import os
import re
from datetime import datetime

from google.oauth2.service_account import Credentials
from googleapiclient.discovery import build

# Constants
SERVICE_ACCOUNT_FILE = "dev_tools/credentials.json"
SPREADSHEET_ID = "1HF6_hLEB8m_v0stp4DLGnIoDjgojvo7fjYz-cysjTMc"
CHANGELONG_SHEET_NAME = "Notes"
END_GAME_RANKINGS_PATH = "src/components/_common/Constants/lastUpdated.ts"

# Authenticate and initialize the Sheets API
SCOPES = ["https://www.googleapis.com/auth/spreadsheets.readonly"]
credentials = Credentials.from_service_account_file(SERVICE_ACCOUNT_FILE, scopes=SCOPES)
service = build("sheets", "v4", credentials=credentials)


def get_changelog_date():
    """Fetch date from Changelog sheet (A2) and return it in MM/DD/YYYY format"""
    result = (
        service.spreadsheets()
        .values()
        .get(spreadsheetId=SPREADSHEET_ID, range=f"{CHANGELONG_SHEET_NAME}!A1")
        .execute()
    )
    date_str = result.get("values", [[None]])[0][0]
    if date_str:
        try:
            date_str = (
                date_str.replace("th,", ",")
                .replace("st,", ",")
                .replace("nd,", ",")
                .replace("rd,", ",")
            )
            date_obj = datetime.strptime(date_str, "%B %d, %Y")
            return date_obj.strftime("%m/%d/%Y")
        except ValueError:
            raise ValueError(f"Invalid date format in Changelog: {date_str}")
    return None


def update_end_game_rankings_update_date(new_date):
    """Update the endGameRankingsUpdateDate in the Constants file"""
    with open(END_GAME_RANKINGS_PATH, "r", encoding="utf-8") as file:
        content = file.read()

    # Use a regex to find the line containing endGameRankingsUpdateDate and replace the date
    updated_content = re.sub(
        r'export const ehpUpdateDate = "\d{2}/\d{2}/\d{4}"',
        f'export const ehpUpdateDate = "{new_date}"',
        content,
    )

    with open(END_GAME_RANKINGS_PATH, "w", encoding="utf-8") as file:
        file.write(updated_content)


# Main logic to process sheets and save as separate JSON files
if __name__ == "__main__":

    changelog_date = get_changelog_date()
    if changelog_date:
        update_end_game_rankings_update_date(changelog_date)
        print(f"Updated ehpUpdateDate to {changelog_date}")
    else:
        print("Changelog date is not available, skipping update.")

File: packages/dev/src/changelog/changelog.ts
import { google } from "googleapis"

const SERVICE_ACCOUNT_FILE = "credentials.json"
const CHANGELOG_PATH = "src/constants/lastUpdated.ts"
const SCOPES = ["https://www.googleapis.com/auth/spreadsheets.readonly"]

const SPREADSHEETS = [
  {
    spreadsheet_id: "1HF6_hLEB8m_v0stp4DLGnIoDjgojvo7fjYz-cysjTMc",
    sheet_name: "Notes",
    cell_range: "A1",
    date_format: "%B %d, %Y",
    update_key: "ehpUpdateDate",
  },
  {
    spreadsheet_id: "13YbPw3dM2eN6hr3YfVABIK9LVuCWnVZF0Zp2BGOZXc0",
    sheet_name: "Changelog",
    cell_range: "A2",
    date_format: "%Y/%m/%d",
    update_key: "endGameRankingsUpdateDate",
  },
] as const

const auth = new google.auth.GoogleAuth({
  keyFile: SERVICE_ACCOUNT_FILE,
  scopes: SCOPES,
})

const sheets = google.sheets({ version: "v4", auth })

const getChangelogDate = async (
  spreadsheet_id: string,
  sheet_name: string,
  cell_range: string,
  date_format: string,
): Promise<string | null> => {
  try {
    const result = await sheets.spreadsheets.values.get({
      spreadsheetId: spreadsheet_id,
      range: `${sheet_name}!${cell_range}`,
    })

    let dateStr = result.data.values?.[0]?.[0]
    if (!dateStr) {
      return null
    }

    if (date_format === "%B %d, %Y") {
      dateStr = dateStr
        .replace("th,", ",")
        .replace("st,", ",")
        .replace("nd,", ",")
        .replace("rd,", ",")
    }

    const dateObj = new Date(dateStr)
    if (isNaN(dateObj.getTime())) {
      throw new Error(`Invalid date format in Changelog: ${dateStr}`)
    }

    const month = String(dateObj.getMonth() + 1).padStart(2, "0")
    const day = String(dateObj.getDate()).padStart(2, "0")
    const year = dateObj.getFullYear()
    return `${month}/${day}/${year}`
  } catch (error) {
    if (error instanceof Error) {
      console.error(`Error fetching sheet: ${error.message}`)
    } else {
      console.error(`An unknown error occurred while fetching sheet`)
    }
    return null
  }
}

const updateConstantsFile = async (updates: Record<string, string>) => {
  let fileContent = ""
  for (const [key, new_date] of Object.entries(updates)) {
    fileContent += `export const ${key} = "${new_date}"\n`
  }

  await Bun.write(CHANGELOG_PATH, fileContent)
}

export const main = async () => {
  const updates: Record<string, string> = {}

  for (const sheetConfig of SPREADSHEETS) {
    const changelogDate = await getChangelogDate(
      sheetConfig.spreadsheet_id,
      sheetConfig.sheet_name,
      sheetConfig.cell_range,
      sheetConfig.date_format,
    )
    if (changelogDate) {
      updates[sheetConfig.update_key] = changelogDate
    }
  }

  if (Object.keys(updates).length > 0) {
    await updateConstantsFile(updates)
    console.log(`Updated dates: ${JSON.stringify(updates)}`)
  } else {
    console.log("No valid dates found, skipping update.")
  }
}

main().catch((err) => {
  console.error("An error occurred", err)
  process.exit(1)
})
